src/shared/Microsoft.Git.CredentialManager/Interop/LibGit2.cs

```diff
@@ -0,0 +1,154 @@
+
+            // If we have a repository path then also include the local repository configuration
+            if (repositoryPath != null)
+            {
+                // We don't need to check for the file's existence since libgit2 will do that for us!
+                string repoConfigPath = Path.Combine(repositoryPath, ".git", "config");
+
+                // Add the repository configuration
+                int error = git_config_add_file_ondisk(config, repoConfigPath, GIT_CONFIG_LEVEL_LOCAL, null, 0);
+                switch (error)
+                {
+                    case GIT_OK:
+                    case GIT_ENOTFOUND: // If the file was not found we should just continue
+                        break;
+                    default:
+                        git_config_free(config);
+                        ThrowIfError(error, nameof(git_config_add_file_ondisk));
+                        return null;
+                }
+            }
+
+            return new LibGit2Configuration(config, repositoryPath);
+        }
+
+        public string GetRepositoryPath(string path)
+        {
+            var buf = new git_buf();
+            int error = git_repository_discover(buf, path, true, null);
+
+            try
+            {
+                switch (error)
+                {
+                    case GIT_OK:
+                        return buf.ToString();
+                    case GIT_ENOTFOUND:
+                        return null;
+                    default:
+                        ThrowIfError(error, nameof(git_repository_discover));
+                        return null;
+                }
+            }
+            finally
+            {
+                git_buf_dispose(buf);
+            }
+        }
+
+        private void Dispose(bool disposing)
+        {
+            git_libgit2_shutdown();
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~LibGit2()
+        {
+            Dispose(false);
+        }
+    }
+
+    internal class LibGit2Configuration : IGitConfiguration
+    {
+        private readonly unsafe git_config* _config;
+        private readonly unsafe git_config* _snapshot;
+
+        internal unsafe LibGit2Configuration(git_config* config, string repositoryPath)
+        {
+            _config = config;
+            RepositoryPath = repositoryPath;
+
+            // Create snapshot for reading values
+            _snapshot = CreateSnapshot(_config);
+        }
+
+        public string RepositoryPath { get; }
+
+        public string this[string name] => GetString(name);
+
+        public unsafe string GetString(string name)
+        {
+            int result = git_config_get_string(out string value, _snapshot, name);
+
+            switch (result)
+            {
+                case GIT_ENOTFOUND:
+                    return null;

```

Comment:
I appreciate that `config["key"]` will not throw a `KeyNotFoundException` like a dictionary would. We'll need to be careful about assuming default values when we see a null or empty on read.

Reply:
I've changed the design here. Instead we now have a `TryGetString` method which is on the interface, and extension methods for `GetString` (and the overloads).